### PR TITLE
Add `dry_run` flag to Produce Records API

### LINKED ISSUE #159: Add message validation functionality to Sidecar
Extend the produce records API to take a `?dryRun=true` query parameter and perform the following validations:

1. Does the schema version exist?
1. Does the specified subject exist? (Depends on #111)
1. Does the key content match the schema (if one is specified)?
1. Does the value content match the schema (if one is specified)?
1. Does the partition exist?

### PR DESCRIPTION
**Pull request by rohitsanj**
## Summary of Changes

Resolves #159 

The API change made in the PR to include a query parameter `dry_run` will not appear in the generated OpenAPI spec. The Internal Kafka REST routes are excluded from the generated OpenAPI spec by design. The query parameter will need to be documented in the [`ce-kafka-rest.openapi.yaml`](https://github.com/confluentinc/vscode/blob/main/src/clients/sidecar-openapi-specs/ce-kafka-rest.openapi.yaml) of confluentinc/vscode. (https://github.com/confluentinc/vscode/issues/613 tracks that work).

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
    - [ ] Updated existing
    - [ ] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

File: src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java [hunk 1]
```diff
@@ -63,18 +65,33 @@ public Uni<ProduceResponse> produceRecord(
...
+          // This should indicate that the request as a whole is valid.
+          .map(c -> ProduceResponse
+              .builder()
+              .clusterId(c.clusterId)
+              .topicName(c.topicName)
+              .build()
+          );
+    } else {
+      return uptoDryRun
+          .chain(this::sendSerializedRecord)
```
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java:92]: Do we need to chain this with `.onFailure(RuntimeException.class).transform(this::unwrapRootCause)`?

**Review by flippingbits — COMMENTED** [review]
LGTM! WDYT of implementing the `dryRun` flag as an enum with the initial values `NONE` and `FULL`? For other endpoints, we might want to offer (partial) dry runs.

**Comment by rohitsanj** [comment]
> WDYT of implementing the dryRun flag as an enum with the initial values NONE and FULL? For other endpoints, we might want to offer (partial) dry runs.

That makes sense! I'd like to keep it simple for now though since we don't need partial dry run for this particular endpoint.

File: src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java (hunk 1 above)
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java:92]: Ah, yes. Thanks

**Comment by flippingbits** [comment]
> I'd like to keep it simple for now though since we don't need partial dry run for this particular endpoint.

👍 Note that changing the query param from boolean to enum is a breaking change. OTOH, the VS Code extension is the only consumer of this API so we should be able to apply the change without causing harm if we communicate it clearly.

**Review by flippingbits — APPROVED** [review]

**Comment by rohitsanj** [comment]
> Note that changing the query param from boolean to enum is a breaking change. OTOH, the VS Code extension is the only consumer of this API so we should be able to apply the change without causing harm if we communicate it clearly.

Thanks! I appreciate the well-reasoned approach.

File: src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java [hunk 2]
```diff
@@ -63,18 +65,35 @@ public Uni<ProduceResponse> produceRecord(
       @HeaderParam(CONNECTION_ID_HEADER) String connectionId,
       @PathParam("cluster_id") String clusterId,
       @PathParam("topic_name") String topicName,
+      @QueryParam("dry_run") @DefaultValue("false") boolean dryRun,
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImpl.java:68]: If we're adding a query parameter, wouldn't the OpenAPI spec change?

**Review by rhauch — DISMISSED** [review]
LGTM

**Review by flippingbits — APPROVED** [review]
LGTM. Thanks, @rohitsanj!

**Pull request merged**